radix-engine/src/vm/wasm_runtime/scrypto_runtime.rs

```diff
@@ -658,4 +683,50 @@ impl<'y, Y: SystemApi<RuntimeError>> WasmRuntime for ScryptoRuntime<'y, Y> {
 
         self.allocate_buffer(hash.to_vec())
     }
+
+    #[trace_resources(log=data.len())]
+    fn crypto_utils_blake2b_256_hash(
+        &mut self,
+        data: Vec<u8>,
+    ) -> Result<Buffer, InvokeError<WasmRuntimeError>> {
+        // TODO: costing
+
+        let hash = blake2b_256_hash(data);
+
+        self.allocate_buffer(hash.to_vec())
+    }
+
+    #[trace_resources(log=message.len())]
+    fn crypto_utils_ed25519_verify(
+        &mut self,
+        message: Vec<u8>,
+        public_key: Vec<u8>,
+        signature: Vec<u8>,
+    ) -> Result<u32, InvokeError<WasmRuntimeError>> {
+        let public_key: Ed25519PublicKey =
+            scrypto_decode(&public_key).map_err(WasmRuntimeError::InvalidPublicKey)?;
```

Comment:
As discussed. SBOR encoding/decoding not really needed for keys and signatures.
However I am using it anyway to not introduce yet another error variant. `InvalidPublicKey` and `InvalidSignature` include `DecodeError`

Reply:
> As discussed. SBOR encoding/decoding not really needed for keys and signatures.
However I am using it anyway to not introduce yet another error variant. InvalidPublicKey and InvalidSignature include DecodeError

Whilst it's not a big deal to SBOR encoding for validation as I assume validation is a lot more compute intensive, I'm still  not sure I follow this argument. Why can't we use `Ed25519PublicKey::try_from(public_key.as_slice())` if `public_key` were a `Vec<u8>`? Or even more efficiently, implement `TryFrom<Vec<u8>> for Ed25519PublicKey`?

Reply:
My point was that, when we use `Ed25519PublicKey::try_from(public_key.as_slice())` we cannot return `WasmRuntime::InvalidPublicKey(DecodeError)` in case of failure.
We will need to add another error `WasmRuntime::InvalidPublicKey(ParseBlsPublicKeyError)` (and similarly for signature).
If we are fine with adding new error variants, then I am happy to do change it :)

Reply:
Yeah that's fine as far as I'm concerned 👍. Just make sure to add new variants rather than changing/breaking existing ones, else the SBOR schema tests will complain at you :D.

Reply:
Got rid of SBOR encode/decode. New error variants added.